ext/openssl/ossl_pkey.c

```diff
@@ -446,6 +446,174 @@ pkey_generate(int argc, VALUE *argv, VALUE self, int genparam)
+                    OSSL_PARAM_BLD_free(params->param_bld);
+                    ossl_raise(ePKeyError, "OSSL_PARAM_BLD_push_utf8_string");
+                }
+                break;
+
+            case OSSL_PARAM_OCTET_STRING:
+                StringValue(value);
+                if(!OSSL_PARAM_BLD_push_octet_string(params->param_bld, key_ptr, RSTRING_PTR(value), RSTRING_LENINT(value))) {
+                    OSSL_PARAM_BLD_free(params->param_bld);
+                    ossl_raise(ePKeyError, "OSSL_PARAM_BLD_push_octet_string");
+                }
+                break;
+            case OSSL_PARAM_UTF8_PTR:
+            case OSSL_PARAM_OCTET_PTR:
+                OSSL_PARAM_BLD_free(params->param_bld);
+                ossl_raise(ePKeyError, "Unsupported parameter \"%s\", handling of OSSL_PARAM_UTF8_PTR and OSSL_PARAM_OCTET_PTR not implemented", key_ptr);
+                break;
+            }
+
+            return ST_CONTINUE;
+        }
+    }
+    OSSL_PARAM_BLD_free(params->param_bld);
+
+    char message_buffer[512] = { 0 };
+    char *cur = message_buffer;
+    char *end = message_buffer + sizeof(message_buffer);
+    for (const OSSL_PARAM *settable_params = params->settable_params; settable_params->key != NULL; settable_params++) {
+        const char *fmt = cur == message_buffer ? "%s" : ", %s";
+        if (cur > end)
+            break;
+        cur += snprintf(cur, end-cur, fmt, settable_params->key);
+    }
+
+    for(int i = 0; strlen(params->aliases[i].alias) > 0; i++) {
+        const char *fmt = cur == message_buffer ? "%s" : ", %s";
+        if (cur > end)
+            break;
+        cur += snprintf(cur, end-cur, fmt, params->aliases[i].alias);
+    }
+
+    ossl_raise(ePKeyError, "Invalid parameter \"%s\". Supported parameters: \"%s\"", key_ptr, message_buffer);
+}
+
+static VALUE
+pkey_from_parameters(int argc, VALUE *argv, VALUE self)
+{
+    VALUE alg, options;
+    rb_scan_args(argc, argv, "11", &alg, &options);
+
+    const char* algorithm = StringValueCStr(alg);
+
+    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_from_name(NULL, algorithm, NULL);
+
+    if (ctx == NULL)
+        ossl_raise(ePKeyError, "EVP_PKEY_CTX_new_from_name");
+
+    struct pkey_from_parameters_arg from_params_args = { 0 };
+
+    from_params_args.param_bld = OSSL_PARAM_BLD_new();
+
+    if (from_params_args.param_bld == NULL) {
+        EVP_PKEY_CTX_free(ctx);
+        ossl_raise(ePKeyError, "OSSL_PARAM_BLD_new");
+    }
+
+    from_params_args.settable_params = EVP_PKEY_fromdata_settable(ctx, EVP_PKEY_KEYPAIR);
+
+    if (from_params_args.settable_params == NULL) {
+        EVP_PKEY_CTX_free(ctx);
+        ossl_raise(ePKeyError, "EVP_PKEY_fromdata_settable");
+    }
+
+    if (strcmp("RSA", algorithm) == 0)
+        from_params_args.aliases = rsa_aliases;
+    else
+        from_params_args.aliases = fcc_aliases;
+
+    rb_hash_foreach(options, &add_parameter_to_builder, (VALUE) &from_params_args);
```

Comment:
`ctx` will leak if `rb_hash_foreach()` raises an exception. Please wrap it with `rb_protect()` so that they can freed before escaping from this function.

Reply:
Adjusted this one. There were more leaks than just this one but now I think it might be freeing everything when exceptions happen. This approach also removes the need to free things inside the iterator function.